Support for multiple Fibaro gateways

pbalogh77 opened the pull request:
## Description:
Fibaro component now supports multiple Fibaro HC gateways.

- Added multiple gateway support.
- Reworked parameter flow to platforms to enable multiple controllers.

**Breaking change:**
A list of gateways is expected instead of a single config.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
fibaro:
    gateways:
      - url: http://your.fibaro.url/api
        username: username
        password: password
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io/pull/8140) PR 8140

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

pbalogh77 committed:
Preparing for transition to config flow

Added multiple gateway support
Reworked parameter flow to platforms to enable multiple controllers
Breaking change to config, now a list of gateways is expected instead of a single config

pbalogh77 committed:
Updated coveragerc

Added new location of fibaro component

dgomes reviewed (approved):
Looks good!

balloob commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -77,18 +78,28 @@
 
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_GATEWAYS, default={}):
```
default should be a list too?

balloob replied:
Now this will be converted to be `[{}]`, if that is expected, make that the default.

balloob commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -65,7 +66,7 @@
 
-CONFIG_SCHEMA = vol.Schema({
+GATEWAY_CONFIG = vol.Schema({
     DOMAIN: vol.Schema({
```
Why keep DOMAIN in GATEWAY_CONFIG ?

pbalogh77 replied:
I've removed it and it seems to work just as well ;-)

balloob commented:
ok to merge when comments addressed

pbalogh77 committed:
Fixes based on code review and extended logging

Addressed issues raised by code review
Added extended debug logging to get better reports from users if the device type mapping is not perfect

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -242,33 +253,44 @@ def _read_devices(self):
-                         config[DOMAIN][CONF_PLUGINS],
-                         config[DOMAIN])
+    config = base_config.get(DOMAIN)
+    gateways = config.get(CONF_GATEWAYS)
```
This looks bad. Don't use `dict.get` twice in a row on consecutive nested keys without default. Not knowing about the config validation, it looks like if the first key would be missing, the first lookup would return `None` and we would get an error here for the second key.

Since we know that those keys will exist in the dict, due to config validation, we don't even need to use `dict.get`. Just do `dict[key]`.

pbalogh77 replied:
Thanks, I fixed it

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -65,30 +66,38 @@
         from fiblary3.client.v4.client import Client as FibaroClient
-        self._client = FibaroClient(url, username, password)
+
+        self._client = FibaroClient(config.get(CONF_URL),
```
Don't use `dict.get` for required config keys or keys with default config schema values. Use `dict[key]`.

pbalogh77 replied:
Thanks, fixed

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -242,33 +253,44 @@ def _read_devices(self):
-        hass.data[FIBARO_DEVICES] = controller.fibaro_devices
+        for controller in hass.data[FIBARO_CONTROLLERS].values():
+            controller.disable_state_handler()
+        hass.data[FIBARO_CONTROLLERS] = {}
```
If we don't need this key anymore we should pop it or delete it from the dict.

pbalogh77 replied:
Wouldn't that be a less efficient algorithm, tho?
I know it's highly theoretical either way, since it's usually a single element list and it's probably equally readable either way, but this seems like a more appropriate solution on an embedded device. I could be wrong, I'm new here. Of course if you insist, I'll modify it.

MartinHjelmare replied:
I don't think we should worry about speed here, but clean up remaining references.

pbalogh77 replied:
Ok, I'll keep in mind.

MartinHjelmare replied:
Please remove this line.

pbalogh77 replied:
Ok, I removed it. Just for my education, what was wrong with that line? I'm new to python but as far as I understand it did remove the reference to all items in the map and by that freeing up the memory. I understand your suggestion that you'd prefer enumerate + pop, but coming from decades of embedded c programming and knowing how enumerate and pop works, that's really a very wasteful way of solving the problem. Afaik, enumerate iterates through the list an extra time, creates a new index, and pop unlinks items one-by-one from the list. I could be totally wrong on my assumptions, but in general this construct feels like something to discourage on an embedded platform. I could be totally not pythonic in thinking this.
Anyhow, I'd love to learn. Thanks in advance.

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -65,30 +66,38 @@
-        vol.Optional(CONF_EXCLUDE, default=[]): FIBARO_ID_LIST_SCHEMA,
-        vol.Optional(CONF_DEVICE_CONFIG, default={}):
-            vol.Schema({cv.string: DEVICE_CONFIG_SCHEMA_ENTRY})
+        vol.Optional(CONF_GATEWAYS, default=[{}]):
```
This default won't validate in the config schema. Why set a default at all?

pbalogh77 replied:
Thanks, fixed

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -65,30 +66,38 @@
         # Whether to import devices from plugins
-        self._import_plugins = import_plugins
-        self._device_config = config[CONF_DEVICE_CONFIG]
+        self._import_plugins = config.get(CONF_PLUGINS, False)
```
The default is already present in the config validation schema.

pbalogh77 replied:
Thanks, fixed

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -65,30 +66,38 @@
-        self._import_plugins = import_plugins
-        self._device_config = config[CONF_DEVICE_CONFIG]
+        self._import_plugins = config.get(CONF_PLUGINS, False)
+        self._device_config = config.get(CONF_DEVICE_CONFIG, {})
```
See above.

pbalogh77 replied:
Thanks, fixed

pbalogh77 committed:
Changhes based on code review

Changes to how configuration is read and schemas
Fix to device type mapping logic

pbalogh77 committed:
simplified reading config

pbalogh77 committed:
oops

oops

pbalogh77 committed:
grr

grr

pbalogh77 committed:
change based on code review

pbalogh77 commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -242,33 +252,44 @@ def _read_devices(self):
-            discovery.load_platform(hass, component, DOMAIN, {}, config)
-        controller.enable_state_handler()
+            discovery.load_platform(hass, component, DOMAIN, {},
+                                    base_config[DOMAIN])
```
ok, I changed it, thanks. AFAIK it's never used, tho.

pbalogh77 committed:
changes based on code review

changes based on code review

MartinHjelmare reviewed (approved):
Good!

MartinHjelmare commented:
Can be merged when docs PR is linked in the PR description.

pbalogh77 commented:
Updated documentation, added link to PR
